p2p/messenger.go

```diff
@@ -0,0 +1,342 @@
+package p2p
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/centrifuge/go-centrifuge/errors"
+
+	"github.com/libp2p/go-libp2p-host"
+	"github.com/libp2p/go-libp2p-protocol"
+
+	pb "github.com/centrifuge/go-centrifuge/protobufs/gen/go/protocol"
+	ggio "github.com/gogo/protobuf/io"
+	"github.com/jbenet/go-context/io"
+	inet "github.com/libp2p/go-libp2p-net"
+	"github.com/libp2p/go-libp2p-peer"
+)
+
+// ErrReadTimeout must be used when receiving timeout while reading
+const ErrReadTimeout = errors.Error("timed out reading response")
+
+// ErrInvalidatedMessageSender must be used when the message sender object created is no longer valid (connection has dropped)
+const ErrInvalidatedMessageSender = errors.Error("message sender has been invalidated")
+
+type bufferedWriteCloser interface {
+	ggio.WriteCloser
+	Flush() error
+}
+
+// The Protobuf writer performs multiple small writes when writing a message.
+// We need to buffer those writes, to make sure that we're not sending a new
+// packet for every single write.
+type bufferedDelimitedWriter struct {
+	*bufio.Writer
+	ggio.WriteCloser
+}
+
+func newBufferedDelimitedWriter(str io.Writer) bufferedWriteCloser {
+	w := bufio.NewWriter(str)
+	return &bufferedDelimitedWriter{
+		Writer:      w,
+		WriteCloser: ggio.NewDelimitedWriter(w),
+	}
+}
+
+func (w *bufferedDelimitedWriter) Flush() error {
+	return w.Writer.Flush()
+}
+
+type p2pMessenger struct {
+	host host.Host // the network services we need
+	self peer.ID   // Local peer (yourself)
+
+	timout time.Duration
+	ctx    context.Context
+
+	strmap map[peer.ID]map[protocol.ID]*messageSender
+	smlk   sync.Mutex
+
+	plk sync.Mutex
+
+	msgHandlers map[pb.MessageType]func(ctx context.Context, peer peer.ID, protoc protocol.ID, msg *pb.P2PEnvelope) (*pb.P2PEnvelope, error)
```

Comment:
Lets go with simple handler instead. I can see all handlers having the same boilerplate code of marshalling and unmarshalling. Instead, this can be done by p2p with all marshaling and unmarshalling

Reply:
Just talked to Ved, to clarify I think this makes sense:

Unmarshal the embedded message in `P2PEnvelope` to the right protobuf message and pass it in. This would remove a lot of redundant code for unmarshaling & error checking and would allow the handler to just do one type assertion.

```suggestion
 	msgHandlers map[pb.MessageType]func(ctx context.Context, peer peer.ID, protoc protocol.ID, msg *proto.Message) (*pb.P2PEnvelope, error)
```

Reply:
What I'm suggesting is not centrifuge specific but rather generic one as we did for Repository interface. Once we integrate this and we decide to make it a library and add the boiler plate code to the p2p itself, we have to refactor all the handlers which can be completely avoided IMO

Reply:
I clarified my comment above with a diagram @vedhavyas @lucasvo since you guys don't seem to have understood what I said.